### ISSUE TITLE
Copy over relevant parts of the `Analytics` class into `Analytics_4`

### ISSUE THREAD

**Issue by nfmohit** [google/site-kit-wp#7926]
## Feature Description

In an effort to use `analytics-4` as the singular Analytics module, the goal is to make the `Analytics` class obsolete and only use the `Analytics_4` class going forward.

Even though most of the `Analytics` class is no longer relevant after the removal of UA, there are still some parts that are still relevant, and thus should be copied over to `Analytics_4`. Some notable pieces are:

1. `handle_provisioning_callback` method: This method co-exists between both `Analytics` and `Analytics_4` classes. They should be merged into one removing the need to use a hook to trigger the other.
2. `get_debug_fields` method: Since the `accountID` setting will be a part of `Analytics_4`, its `get_debug_fields` method should be updated to include debug information for this setting.
3. `register` method: This method includes different callbacks to hooks that should be carefully evaluated and merged.
4. `is_tracking_disabled` & `print_tracking_opt_out` methods: These methods implement the Analytics tracking opt-out for certain user groups. They should be merged accordingly.
5. `Analytics\Advanced_Tracking` class: This class should be migrated to the `Analytics_4` namespace, but not be instantiated, as it will be updated to work with GA4 as a part of [#7145](https://github.com/google/site-kit-wp/issues/7145).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation Brief

* 

### Test Coverage

* 

## QA Brief

* 

## Changelog entry

*

**Comment by tofumatt** [comment]
@nfmohit Because this deals with functionality after the removal of UA, but doesn't explicitly specify what that is in the AC nor is blocked by a UA removal issue, can you add the points you mention in the issue description to the ACs? Just so it's extra-clear what should be moved over and what shouldn't be 🙂 

Then this should be good-to-go 👍🏻

**Comment by nfmohit** [comment]
> @nfmohit Because this deals with functionality after the removal of UA, but doesn't explicitly specify what that is in the AC nor is blocked by a UA removal issue, can you add the points you mention in the issue description to the ACs? Just so it's extra-clear what should be moved over and what shouldn't be 🙂
> 
> Then this should be good-to-go 👍🏻

Thank you for the kind review, @tofumatt. I intentionally wanted to leave this out as open-ended so that the specific discovery could be a part of the IB process. There might be other small pieces still needed outside of the methods/pieces that are mentioned in the issue description which might've been missed while authoring the design doc. According to your feedback, I have included them in the ACs, but also left another point about being cautious of any other piece that might still be needed. LMKWYT, thanks!

**Comment by tofumatt** [comment]
Sounds good, that makes sense 👍🏻 

Moving to IB 🙂

**Comment by techanvil** [comment]
Hey @zutigrm, thanks for drafting this IB. A few notes:

- The points about `googlesitekit_analytics_adsense_linked` and `exception_to_error()` are not needed as these relate to the UA AdSense linking functionality which is no longer relevant. GA4 does now provide the ability to link to AdSense but works a bit differently, and we have an [epic](https://app.asana.com/0/1206011557464713/1206011352527238) underway to add support for it.
- The `googlesitekit_dashboard_sharing_data` filter hook in `Analytics` doesn't actually do anything useful - in fact, it's clear we missed this aspect in this recent [PR](https://github.com/google/site-kit-wp/pull/7581/files#diff-40a2ef802e3a439ddd0f6c602524bf2ade1f5929b6dae1d4588894a60e95ecd9) and the `add_filter()` call should have been removed entirely. Seeing as the `Analytics` class is being removed anyway, we can leave it there but there's no need to copy it to `Analytics_4`.
- I think we'll still need the `GET:accounts-properties-profiles` endpoint until we fully move from the analytics `getAccounts()` to `getAccountSummaries()`, please could you check this and update as necessary?
- We'll also need to migrate the check for `accountID` in `is_connected()`.
- Lastly, it's by no means essential, but you could update the first two permalinks as the linked code has changed slightly (or at least the surrounding context of it has).

**Comment by zutigrm** [comment]
@techanvil Thank you for the review.

> exception_to_error

This seems to be needed, as it is used in Analytics 4 method [here](https://github.com/google/site-kit-wp/blob/7b715060ca674fdbdb0f1d753b8306af59498613/includes/Modules/Analytics_4.php#L1617)

> I think we'll still need the GET:accounts-properties-profiles endpoint until we fully move from the analytics getAccounts() to analytics-4 getAccountSummaries(), please could you check this and update as necessary?

This is being done as part of #7637, which is switching to account summaries, so `GET:accounts-properties-profiles` doesn't seem to be needed

> We'll also need to migrate the check for accountID in is_connected().

This will be done as part of #7932 where final migration and removal of old analytics module is done, which also turns off the internal property and makes it only Analytics module, which makes more sense that is_connected will be checked against accountID, but I can include it here is well, if it makes sense to be done as part of this issue?

Let me know what you think

**Comment by techanvil** [comment]
Thanks @zutigrm.

> exception_to_error
> 
> This seems to be needed, as it is used in Analytics 4 method [here](https://github.com/google/site-kit-wp/blob/7b715060ca674fdbdb0f1d753b8306af59498613/includes/Modules/Analytics_4.php#L1617)

This `Analytics_4` method calls the version of `exception_to_error()` defined in the [superclass](https://github.com/google/site-kit-wp/blob/7b715060ca674fdbdb0f1d753b8306af59498613/includes/Core/Modules/Module.php#L644), it doesn't need the overridden version that `Analytics` provides.

> > I think we'll still need the GET:accounts-properties-profiles endpoint until we fully move from the analytics getAccounts() to analytics-4 getAccountSummaries(), please could you check this and update as necessary?
> 
> This is being done as part of #7637, which is switching to account summaries, so `GET:accounts-properties-profiles` doesn't seem to be needed

Good spot there, however, if you review the search results for Analytics `getAccounts()` vs the IB for #7637, there will still be some remaining usage of it. As the description for that issue alludes to, it's just a start of the migration from `getAccounts()` to `getAccountSummaries()` so we'll still need this in the short term at least.

> > We'll also need to migrate the check for accountID in is_connected().
> 
> This will be done as part of #7932 where final migration and removal of old analytics module is done, which also turns off the internal property and makes it only Analytics module, which makes more sense that is_connected will be checked against accountID, but I can include it here is well, if it makes sense to be done as part of this issue?

Good spot, and seeing as the change to `is_connected()` is specced in the IB for that issue, we can leave it out of this one.

**Comment by zutigrm** [comment]
@techanvil Got it, thanks! IB updated

**Comment by techanvil** [comment]
Thanks @zutigrm! IB LGTM. :white_check_mark:

**Comment by techanvil** [comment]
Hey @jimmymadon @zutigrm - apologies for missing this during the IB review, but I've just realised we are creating a GA4 version of `getAccounts()` in https://github.com/google/site-kit-wp/issues/7637, which uses `accountSummaries` under the hood, so we should in fact no longer need the `GET:accounts-properties-profiles` endpoint after all.

**Comment by nfmohit** [comment]
@aaemnnosttv @jimmymadon I have added the following to the ACs for clarification:

> * If a piece copied to `Modules\Analytics_4` retains its functionality even without being present in the `Modules\Analytics` class, and if it is deemed safe to remove it from `Modules\Analytics`, it should be eliminated.
> * It should be ensured that no duplication is introduced due to the migration here, e.g. duplicate tags, duplicate hooks, etc.

Please let me know if that looks good, thank you!

**Comment by mohitwp** [comment]
### QA Update ✅ 

- Tested on dev environment.
- Created new GA4 a/c on fresh site.
- Verified all Ga4 settings under edit component.
- Verified that all fields are getting save as per selection.
- Verified that user is able to create new property.
- Verified that user is able to disconnect and connect analytics.
- Verified analytics functionality with latest environment.
- Verified analytics Opt Out Snippet.
- Verified analytics snippet.
- Also, verified snippet on site on which UA property is connected by switching on to dev environment.
- Analytics and opt out snippet is same as latest environment.

<details>

GA4 snippet with Key metrics

![image](https://github.com/google/site-kit-wp/assets/94359491/088657a0-a69d-4436-8354-4c6a2b830dfd)

Ga4 snippet with Ads conversion

![image](https://github.com/google/site-kit-wp/assets/94359491/ae9a2d32-1b63-4c30-8489-4d5e576ad384)

GA4 opt out snippet

![image](https://github.com/google/site-kit-wp/assets/94359491/d8137d4d-3c01-40f4-bf9e-376b1fb9aec0)

</details>

**Comment by mohitwp** [comment]
### QA Update ✅ 

Thanks @jimmymadon !

- Tested on dev environment.
- Created new GA4 a/c on fresh site.
- Verified all Ga4 settings under edit component.
- Verified that all fields are getting save as per selection.
- Verified that user is able to create new property.
- Verified that user is able to disconnect and connect analytics.
- Verified analytics functionality with latest environment.
- Verified analytics Opt Out Snippet.
- Verified analytics snippet.
- Also, verified snippet on site on which UA property is connected by switching on to dev environment.
- Analytics and opt out snippet is same as latest environment.

<details>

GA4 snippet with Key metrics

![image](https://github.com/google/site-kit-wp/assets/94359491/088657a0-a69d-4436-8354-4c6a2b830dfd)

Ga4 snippet with Ads conversion

![image](https://github.com/google/site-kit-wp/assets/94359491/ae9a2d32-1b63-4c30-8489-4d5e576ad384)

GA4 opt out snippet

![image](https://github.com/google/site-kit-wp/assets/94359491/d8137d4d-3c01-40f4-bf9e-376b1fb9aec0)

</details>

**Comment by bethanylang** [comment]
@mohitwp Should this now be in Approval? Thanks!

**Comment by mohitwp** [comment]
@bethanylang It's required `QA:Eng `review.

**Comment by bethanylang** [comment]
@mohitwp Ah, missed that, thank you!
